src/db/spanner/models.rs

```diff
@@ -783,6 +799,172 @@ impl SpannerDb {
+            .one_or_none()
+            .await?;
+        let set_sql = if let Some(result) = result {
+            // Update the user_collections table to reflect current numbers.
+            // If there are BSOs, there are user_collections (or else something
+            // really bad already happened.)
+            if self.quota_enabled {
+                let mut v = Value::new();
+                v.set_string_value(result[0].get_string_value().to_owned());
+                sqlparams.insert("total_bytes".to_owned(), v);
+                let mut v = Value::new();
+                v.set_string_value(result[1].get_string_value().to_owned());
+                sqlparams.insert("count".to_owned(), v);
+                sqltypes.insert(
+                    "total_bytes".to_owned(),
+                    crate::db::spanner::support::as_type(TypeCode::INT64),
+                );
+                sqltypes.insert(
+                    "count".to_owned(),
+                    crate::db::spanner::support::as_type(TypeCode::INT64),
+                );
+                "UPDATE user_collections
+                SET modified = @modified,
+                    count = @count,
+                    total_bytes = @total_bytes
+                WHERE fxa_uid = @fxa_uid
+                    AND fxa_kid = @fxa_kid
+                    AND collection_id = @collection_id"
+            } else {
+                "UPDATE user_collections
+                SET modified = @modified
+                WHERE fxa_uid = @fxa_uid
+                AND fxa_kid = @fxa_kid
+                AND collection_id = @collection_id"
+            }
+        } else {
+            // Otherwise, there are no BSOs that match, check to see if there are
+            // any collections.
+            let result = self
+                .sql(
+                    "SELECT 1 FROM user_collections
+                WHERE fxa_uid=@fxa_uid AND fxa_kid=@fxa_kid AND collection_id=@collection_id
+                LIMIT 1",
+                )?
+                .params(sqlparams.clone())
+                .param_types(sqltypes.clone())
+                .execute_async(&self.conn)?
+                .one_or_none()
+                .await?;
+            if result.is_none() {
+                // No collections, so insert what we've got.
+                if self.quota_enabled {
+                    "INSERT INTO user_collections (fxa_uid, fxa_kid, collection_id, modified, total_bytes, count)
+                    VALUES (@fxa_uid, @fxa_kid, @collection_id, @modified, 0, 0)"
+                } else {
+                    "INSERT INTO user_collections (fxa_uid, fxa_kid, collection_id, modified)
+                    VALUES (@fxa_uid, @fxa_kid, @collection_id, @modified)"
+                }
+            } else {
+                // there are collections, best modify what's there.
+                // NOTE, tombstone is a single collection id, it would have been created above.
+                if self.quota_enabled {
```

Comment:
I'll argue we don't need to set total_bytes/count here because it should already be accurate -- simplifying this into the same update whether or not quota's enabled

Reply:
Kinda depends on paranoia about whether or not things default to 0 correctly, I suppose. You could make the same argument about the `INSERT` call above it. I prefer to be explicit in these cases.